HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

shamser opened the pull request:
Replace this==NULL test with call to static member function that returns pointer if this valid, otherwise returns null.  Remove flag to disable warning regarding this==null check

shamser committed:
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

Replace this==NULL test with call to static member function that returns
pointer if this valid, otherwise returns null.

shamser committed:
:HPCC-13496 'this' pointer cannot be null in well-defined C++ code

Replace this==NULL test with call to static member function that returns
pointer if this valid, otherwise returns null.

shamser committed:
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

Replace this==NULL test with call to static member function that returns
pointer if this valid, otherwise returns null.

shamser committed:
Merge branch 'issue13496' of https://github.com/shamser/HPCC-Platform into issue13496

shamser committed:
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

Remove flag to disable this==null compiler error

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

richardkchapman commented on cmake_modules/commonSetup.cmake:
```diff
@@ -259,11 +259,8 @@ IF ("${COMMONSETUP_DONE}" STREQUAL "")
+      SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}  -Werror=bitwise-op-parentheses ")
       SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}  -Wno-switch-enum -Wno-format-zero-length -Wno-switch")
-      if (CLANG_VERSION VERSION_GREATER 3.6 OR CLANG_VERSION VERSION_EQUAL 3.6)
-        SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-undefined-bool-conversion -Wno-pointer-bool-conversion -Wno-tautological-compare")
```
At least one of these is still needed, as you have not yet fixed the comparisons from &reference to NULL, so we will get build errors on newer versions of Clang

richardkchapman commented on system/jlib/jhash.hpp:
```diff
@@ -49,27 +49,30 @@ interface jlib_decl IAtom : extends IMapping
 {
  public:
     virtual const char * getNamePtr() const = 0;
+    const char *         getAtomNamePtr() const { return IAtom::getAtomNamePtr(this);} // ok if this=NULL
```
In effect you are fooling the compiler again - this is still NULL in this function, but it is no longer detecting the fact because of the call to a static member.

As discussed elsewhere, I think we need to fix properly rather than just hiding the error from the compiler.

At the same time I'd be inclined to rationalize the interface to remove the 3 different synonyms.

@ghalliday What are your thoughts?

ghalliday replied:
I would be very happy to remove the duplicate functions, probably sticking with str() for consistency with other classes.

Richard you are correct, it is still technically invalid and a future iteration of the compiler is likely to catch this and report the problem again.

I think the correct solution is to add a new global function e.g.,

inline const char * str(const IAtom * atom) { return atom ? atom->getAtomName() : NULL; }

And change everything to call that.  Unfortunately I can't think of any change that doesn't require large numbers of changes throughout the source code.

If this change is made to master then we should also add the same global function to 5.4, (or even 5.2.6) so that any new code that is written in those branches can call that new function, and not cause problems when it is up-merged.

I'm not sure about the cast operator.  I think it would be fine to keep it, but it should only be used if the IAtom * is not NULL.  That would probably require removing it, checking all places that call it and modifying any where it could be NULL, and then adding it back again.  It may be much simpler, and safer, to remove it completely.

ghalliday commented:
@richardkchapman do you agree?  @shamser back to you.

richardkchapman commented:
Yes I agree. I would probably remove the cast operator, on the basis of making sure people don't make assumptions in the future. You could argue it's no more likely to cause future errors than (for example) passing a null pointer to a function expecting a const char *, I suppose.

My guess would be that there are not THAT many calls to getAtomNamePtr() that it's going to be a real pain to make the proper change.

shamser commented:
Agree - fixing compiler flags 
Agree - removing synonyms
Agree - removing the const char * cast operator 

This static member function is clearly valid.  It is not only valid for the function to check atom for nullness but it is good practice:
`static const char * getAtomNamePtr(const IAtom *atom) { return atom ? atom->getNamePtr(): NULL; } `

Now we do always call this static member function from within a IAtom Class and always pass 'this':
`const char * getAtomNamePtr() const { return IAtom::getAtomNamePtr(this);}`

Although we are only calling  IAtom::getAtomNamePtr() static member function with a 'this' pointer, that may be not always be the case: as far as the compiler is concerned, the function can be called with any IAtom *. 

So, what is the compiler going to complain about: the definition of the static member function or the call to the static member function?  Is there anything in the standards that say, don't pass 'this' to static member functions?   (Checking nullness of 'this' is very different from passing a 'this' elsewhere to be checked for nullness).

Anyway, I am happy to proceed with the suggested solution - it should achieve the same end result.  @richardkchapman @ghalliday Would you like the code changed to use the global str() solution?

richardkchapman commented:
Yes please.

The point is that in the call

const char * getAtomNamePtr() const { return IAtom::getAtomNamePtr(this);}

the value of this may be NULL. A clever compiler that decided to inline IAtom::getAtomNamePtr could generate the same warning as we currently get - but that's really less of a concern than that (in some hypothetical future architecture we may target) calling a non-static member function with a NULL this pointer may crash (the behaviour is undefined).

We currently rely on this undefined behaviour and assume that so long as we don't access any member variables when this is null, we are ok. But it's not safe or sensible to rely on it, and THAT is what we are really wanting to fix, not just suppress the warning.

shamser commented:
@richardkchapman Got it...  I'll implement the global str() access function.

shamser committed:
Reverting changes.. start again

shamser committed:
Merge branch 'master' of https://github.com/hpcc-systems/HPCC-Platform into issue13496

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

shamser committed:
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

- Set tautological-compare compiler flag to be an error so 'this==null' comparision is treated as errors
- tautlogical-compare also treats &reference==null as errors also
- Code to remove &reference==null check by either removing the check where it is unnecessary or converting the reference to a pointer
- Change details:
-- In dali/dfulplus/dfuplus.cpp: void CDfuPlusHelper::exc(const IMultiException& excep,const char *title)

   &except check for nullness - REMOVED

   The argument excep orginate from a calll to CSoapResponseBinding::getExceptions().  This function returns the address of the member function exceptions_, which is initialized by the constructor so should never be null.

-- In ecl/eclplus/DeleteHelper.cpp b/ecl/eclplus/DeleteHelper.cpp: bool AbortHelper::doit(FILE * fp) AND ecl/eclplus/DumpHelper.cpp: book DumpHelper::doit(FILE * fp)

   (&inforesp->getWorkunit() == NULL) - REMOVED

   IClientWUInfoResponse::getWorkunit() is (IConstECLWorkunit &) m_Workunit.getValue()

   and SoapStruct<CECLWorkunit, IConstECLWorkunit> m_Workunit;

   And Soapstruct::getValue() can never be null as return value is a member variable of type CECLWorkUnit.

- In esp/bindings/SOAP/client/soapclient.cpp: int CSoapClient::postRequest(const char* contenttype, const char* soapaction...

   Add check for nullness of soap_request.get and soap_response.get() before calling CHttpRequest::postRequest() which no longer checks the reference for nullness

- In esp/bindings/http/client/httpclient.cpp: int CHttpClient::postRequest(ISoapMessage &req, ISoapMessage& resp)

  Remove check of &req and &resp full nullness as all code modified to ensure all callers conduct the check before calling this function

- In esp/bindings/http/platform/httptransport.cpp: int CHttpMessage::close()

  &m_socket != NULL check removed as m_socket is assigned a the value by constructor and is used by all other class members, m_socket will not be null and check not necessary

- In esp/platform/sechandler.cpp: bool SecHandler::authorizeSecFeature(const char * pszFeatureUrl, const char* Us..

  Add check of pSecondarUser.get() for nullness so that its reference may be passed to m_setMgr->initUser()

- In /esp/services/ws_access/ws_accessService.cpp: bool Cws_accessEx::onAddUser(IEspContext &context, IEspAddUserRequest &req, IEsp...

  Check user.get() for nullness before passing its reference to secmgr->addUser

- In esp/services/ws_fs/ws_fsService.cpp: static void DeepAssign(IEspContext &context, IConstDFUWorkUnit *src, IEspDFUWork

  Check for &dest==NULL - REMOVED

  The 'dest' argument will never be null as all callers originate from CGetDFUWorkunitResponse::updateResult() and in all cases this returns the address of a member variable m_result.

- In system/security/LdapSecurity/aci.cpp: CSecurityDescriptor* AciProcessor::createDefaultSD(ISecUser * const user

  First argument changed from reference to pointer so that it is clear that the pointer may be null and appropriate check is carried out for nullness

- system/security/LdapSecurity/ldapsecurity.cpp: bool CLdapSecManager::addUser(ISecUser & user)

  &user == NULL check - REMOVED

  All callers have been checked to ensure null pointers are not being passed/cannot be passed

- system/security/shared/caching.cpp: ISecUser* CPermissionsCache::getCachedUser( ISecUser& sec_user)

  &sec_user == NULL check - REMOVED

  All callers have been checked to ensure null pointers are not being passed/cannot be passed

- Remove unnecessary check of IPropertyTree & for nullness that was returned in IPropertyTreeIterator.
ForEach() calls IPropertyTreeIterator's isValid member function to ensure item is vaid.

- IAtom: Remove multiple aliases to access getNamePtr() from IAtom
- Add global str(IAtom *) function to safely access getNamePtr() even if IAtom * argument is null
- IIdAtom: Remove multiple aliases to access getNamePtr() from IAtom
- Add global str(IIdAtom *) function to safely access getNamePtr() even if IIdAtom * argument is null
- Replace all usages of the aliases removed previously with a call to the newly create str() function
- IIdAtom: remove the lower() member variable and replace with global lower(IIdAtom *p) function
- Modify cmake commonSetup.cmake to have this==null treated as compilation errors by CLang compilers

- Remove checks of reference variable’s address
- secuser pointer checked for nullness before passing in argument as reference (working.cpp: CSecureConstWUIterator::checkScope)
- user pointer checked for nullness before passing in argument as reference (ws_DfuService.cpp: getFilePermissions)

shamser committed:
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

- Set tautological-compare compiler flag to be an error so 'this==null' comparision is treated as errors
- tautlogical-compare also treats &reference==null as errors also
- Code to remove &reference==null check by either removing the check where it is unnecessary or converting the reference to a pointer
- Change details:
-- In dali/dfulplus/dfuplus.cpp: void CDfuPlusHelper::exc(const IMultiException& excep,const char *title)

   &except check for nullness - REMOVED

   The argument excep orginate from a calll to CSoapResponseBinding::getExceptions().  This function returns the address of the member function exceptions_, which is initialized by the constructor so should never be null.

-- In ecl/eclplus/DeleteHelper.cpp b/ecl/eclplus/DeleteHelper.cpp: bool AbortHelper::doit(FILE * fp) AND ecl/eclplus/DumpHelper.cpp: book DumpHelper::doit(FILE * fp)

   (&inforesp->getWorkunit() == NULL) - REMOVED

   IClientWUInfoResponse::getWorkunit() is (IConstECLWorkunit &) m_Workunit.getValue()

   and SoapStruct<CECLWorkunit, IConstECLWorkunit> m_Workunit;

   And Soapstruct::getValue() can never be null as return value is a member variable of type CECLWorkUnit.

- In esp/bindings/SOAP/client/soapclient.cpp: int CSoapClient::postRequest(const char* contenttype, const char* soapaction...

   Add check for nullness of soap_request.get and soap_response.get() before calling CHttpRequest::postRequest() which no longer checks the reference for nullness

- In esp/bindings/http/client/httpclient.cpp: int CHttpClient::postRequest(ISoapMessage &req, ISoapMessage& resp)

  Remove check of &req and &resp full nullness as all code modified to ensure all callers conduct the check before calling this function

- In esp/bindings/http/platform/httptransport.cpp: int CHttpMessage::close()

  &m_socket != NULL check removed as m_socket is assigned a the value by constructor and is used by all other class members, m_socket will not be null and check not necessary

- In esp/platform/sechandler.cpp: bool SecHandler::authorizeSecFeature(const char * pszFeatureUrl, const char* Us..

  Add check of pSecondarUser.get() for nullness so that its reference may be passed to m_setMgr->initUser()

- In /esp/services/ws_access/ws_accessService.cpp: bool Cws_accessEx::onAddUser(IEspContext &context, IEspAddUserRequest &req, IEsp...

  Check user.get() for nullness before passing its reference to secmgr->addUser

- In esp/services/ws_fs/ws_fsService.cpp: static void DeepAssign(IEspContext &context, IConstDFUWorkUnit *src, IEspDFUWork

  Check for &dest==NULL - REMOVED

  The 'dest' argument will never be null as all callers originate from CGetDFUWorkunitResponse::updateResult() and in all cases this returns the address of a member variable m_result.

- In system/security/LdapSecurity/aci.cpp: CSecurityDescriptor* AciProcessor::createDefaultSD(ISecUser * const user

  First argument changed from reference to pointer so that it is clear that the pointer may be null and appropriate check is carried out for nullness

- system/security/LdapSecurity/ldapsecurity.cpp: bool CLdapSecManager::addUser(ISecUser & user)

  &user == NULL check - REMOVED

  All callers have been checked to ensure null pointers are not being passed/cannot be passed

- system/security/shared/caching.cpp: ISecUser* CPermissionsCache::getCachedUser( ISecUser& sec_user)

  &sec_user == NULL check - REMOVED

  All callers have been checked to ensure null pointers are not being passed/cannot be passed

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

shamser committed:
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

- Remove unnecessary check of IPropertyTree & for nullness that was returned in IPropertyTreeIterator.
ForEach() calls IPropertyTreeIterator's isValid member function to ensure item is vaid.

shamser committed:
HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

- IAtom: Remove multiple aliases to access getNamePtr() from IAtom
- Add global str(IAtom *) function to safely access getNamePtr() even if IAtom * argument is null
- IIdAtom: Remove multiple aliases to access getNamePtr() from IAtom
- Add global str(IIdAtom *) function to safely access getNamePtr() even if IIdAtom * argument is null
- Replace all usages of the aliases removed previously with a call to the newly create str() function
- IIdAtom: remove the lower() member variable and replace with global lower(IIdAtom *p) function
- Modify cmake commonSetup.cmake to have this==null treated as compilation errors by CLang compilers

shamser committed:
HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

- Remove commented out lines from jhash.hpp

HPCCSmoketest commented:
Automated Smoketest Build: failed
Number of Error(s): 7
/home/ati/smoketest/smoketest-7410/HPCC-Platform/plugins/v8embed/v8embed.cpp:75:51: error: ‘struct IAtom’ has no member named ‘str’
/home/ati/smoketest/smoketest-7410/HPCC-Platform/plugins/v8embed/v8embed.cpp:210:71: error: ‘struct IAtom’ has no member named ‘str’
/home/ati/smoketest/smoketest-7410/HPCC-Platform/plugins/v8embed/v8embed.cpp:213:82: error: ‘struct IAtom’ has no member named ‘str’
/home/ati/smoketest/smoketest-7410/HPCC-Platform/plugins/v8embed/v8embed.cpp:372:51: error: ‘struct IAtom’ has no member named ‘str’
make[2]: *** [plugins/v8embed/CMakeFiles/v8embed.dir/v8embed.cpp.o] Error 1
make[1]: *** [plugins/v8embed/CMakeFiles/v8embed.dir/all] Error 2
make: *** [all] Error 2

shamser committed:
HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

Remove checks of reference variable’s address
secuser pointer checked for nullness before passing in argument as reference (working.cpp: CSecureConstWUIterator::checkScope)
user pointer checked for nullness before passing in argument as reference (ws_DfuService.cpp: getFilePermissions)

HPCCSmoketest commented:
Automated Smoketest Build: failed
Number of Error(s): 7
/home/ati/smoketest/smoketest-7410/HPCC-Platform/plugins/v8embed/v8embed.cpp:75:51: error: ‘struct IAtom’ has no member named ‘str’
/home/ati/smoketest/smoketest-7410/HPCC-Platform/plugins/v8embed/v8embed.cpp:210:71: error: ‘struct IAtom’ has no member named ‘str’
/home/ati/smoketest/smoketest-7410/HPCC-Platform/plugins/v8embed/v8embed.cpp:213:82: error: ‘struct IAtom’ has no member named ‘str’
/home/ati/smoketest/smoketest-7410/HPCC-Platform/plugins/v8embed/v8embed.cpp:372:51: error: ‘struct IAtom’ has no member named ‘str’
make[2]: *** [plugins/v8embed/CMakeFiles/v8embed.dir/v8embed.cpp.o] Error 1
make[1]: *** [plugins/v8embed/CMakeFiles/v8embed.dir/all] Error 2
make: *** [all] Error 2

shamser committed:
HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

Modfiy v8embed.cpp to use the global str() function instead of IAtom:str()

richardkchapman commented:
Looks generally good, minor comments only. The earlier commits should really be removed from the branch via a rebase -i (I can do this during a manual merge if you prefer).

I suspect you also need to make a corresponding pull request in the LN repository.

richardkchapman commented:
Should also make a PR against 5.4.0 that adds the str() and lower() functions (so that any NEW code added in 5.4.0 can be written in such a way that it can upmerge easily).

shamser committed:
Merge branch 'issue13496' of https://github.com/shamser/HPCC-Platform into issue13496

shamser commented:
Closing and starting clean.  New pull request: https://github.com/hpcc-systems/HPCC-Platform/pull/7494
